TauAnalysis/MCEmbeddingTools/plugins/CaloCleaner.cc

```diff
@@ -14,6 +14,90 @@ typedef CaloCleaner<HORecHit> HORecHitColCleaner;
+  std::vector<edm::InputTag> inCollections = iConfig.getParameter<std::vector<edm::InputTag>>("oldCollection");
+  for (const auto &inCollection : inCollections) {
+    inputs_[inCollection.instance()] = consumes<RecHitCollection>(inCollection);
+    produces<RecHitCollection>(inCollection.instance());
+  }
+
+  is_preshower_ = iConfig.getUntrackedParameter<bool>("is_preshower", false);
+  edm::ParameterSet parameters = iConfig.getParameter<edm::ParameterSet>("TrackAssociatorParameters");
+  edm::ConsumesCollector iC = consumesCollector();
+  parameters_.loadParameters(parameters, iC);
+  // trackAssociator_.useDefaultPropagator();
+}
+
+template <typename T>
+CaloCleaner<T>::~CaloCleaner() {
+  // nothing to be done yet...
+}
+
+template <typename T>
+void CaloCleaner<T>::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
+  auto const &propagator = iSetup.getData(propagatorToken_);
+  trackAssociator_.setPropagator(&propagator);
+
+  edm::Handle<edm::View<pat::Muon>> muonHandle;
+  iEvent.getByToken(mu_input_, muonHandle);
+  edm::View<pat::Muon> muons = *muonHandle;
+
+  std::map<uint32_t, float> correction_map;
+
+  // Fill the correction map
+  for (edm::View<pat::Muon>::const_iterator iMuon = muons.begin(); iMuon != muons.end(); ++iMuon) {
+    // get the basic informaiton like fill reco mouon does
+    //     RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc
+    const reco::Track *track = nullptr;
+    if (iMuon->track().isNonnull())
+      track = iMuon->track().get();
+    else if (iMuon->standAloneMuon().isNonnull())
+      track = iMuon->standAloneMuon().get();
+    else
+      throw cms::Exception("FatalError")
+          << "Failed to fill muon id information for a muon with undefined references to tracks";
+    TrackDetMatchInfo info =
+        trackAssociator_.associate(iEvent, iSetup, *track, parameters_, TrackDetectorAssociator::Any);
+    fill_correction_map(&info, &correction_map);
+  }
+
+  // Copy the old collection and correct if necessary
+  for (auto input_ : inputs_) {
+    std::unique_ptr<RecHitCollection> recHitCollection_output(new RecHitCollection());
+    edm::Handle<RecHitCollection> recHitCollection;
+    // iEvent.getByToken(input_.second[0], recHitCollection);
+    iEvent.getByToken(input_.second, recHitCollection);
+    for (typename RecHitCollection::const_iterator recHit = recHitCollection->begin();
+         recHit != recHitCollection->end();
+         ++recHit) {
+      if (correction_map[recHit->detid().rawId()] > 0) {
+        float new_energy = recHit->energy() - correction_map[recHit->detid().rawId()];
+        if (new_energy <= 0)
+          continue;  // Do not save empty Hits
+        T newRecHit(*recHit);
+        newRecHit.setEnergy(new_energy);
+        recHitCollection_output->push_back(newRecHit);
+      } else {
+        recHitCollection_output->push_back(*recHit);
+      }
+      /* For the inveted collection
+           if (correction_map[recHit->detid().rawId()] > 0){
```

Comment:
please remove all commented code, unless really needed

Reply:
Done in a8eae9f